crates/huub/src/constraints/int_value_precede.rs

```diff
@@ -0,0 +1,1144 @@
+				})?;
+			}
+			if i as IntVal <= actions.get_trailed_int(self.last[low as usize])
+				&& actions.check_int_in_domain(v, low)
+			{
+				let _ = actions.set_trailed_int(self.last[low as usize], i as IntVal);
+				low -= 1;
+			}
+			if low == 0 {
+				break;
+			}
+		}
+
+		self.initialized = true;
+		Ok(())
+	}
+
+	/// Iteratively repairs the lower bounds starting with k, only iterates as far
+	/// as necessary.
+	fn repair_lower<P: PropagationActions>(
+		&self,
+		actions: &mut P,
+		mut k: IntVal,
+	) -> Result<(IntVal, IntVal), Conflict> {
+		let mut i = actions.get_trailed_int(self.last[k as usize]);
+		while k > 0 {
+			if actions.check_int_in_domain(self.vars[i as usize], k) {
+				let _ = actions.set_trailed_int(self.last[k as usize], i);
+				if actions.get_trailed_int(self.first[k as usize]) == i {
+					actions.set_int_lower_bound(self.vars[i as usize], k, |a: &mut P| {
+						self.explain_lower(a, i as usize, k)
+					})?;
+				}
+				k -= 1;
+				if actions.get_trailed_int(self.last[k as usize]) < i {
+					return Ok((i, k + 1));
+				}
+			}
+
+			i -= 1;
+			if i < 0 {
+				actions.set_int_lower_bound(self.vars[0], k, |a: &mut P| {
+					self.explain_lower(a, 0, k)
+				})?;
+			}
+		}
+
+		Ok((i, 0))
+	}
+
+	/// Iteratively repair the upper bounds starting with k, only iterates as far
+	/// as necessary.
+	fn repair_upper<P: PropagationActions>(
+		&self,
+		actions: &mut P,
+		mut k: IntVal,
+	) -> Result<(), Conflict> {
+		let mut i = actions.get_trailed_int(self.first[k as usize]);
+		let mut lim = self.get_upper_limit(actions, k as usize);
+
+		while i <= lim {
+			if actions.get_int_upper_bound(self.vars[i as usize]) > k {
+				actions.set_int_upper_bound(self.vars[i as usize], k, |a: &mut P| {
+					self.explain_upper(a, i as usize, k)
+				})?;
+			}
+			if actions.check_int_in_domain(self.vars[i as usize], k) {
+				let _ = actions.set_trailed_int(self.first[k as usize], i);
+				let _ = actions.set_trailed_int(self.first_val[i as usize], k);
+				if actions.get_trailed_int(self.last[k as usize]) == i {
+					actions.set_int_lower_bound(self.vars[i as usize], k, |a: &mut P| {
+						self.explain_lower(a, i as usize, k)
+					})?;
+				}
+				k += 1;
+				if (k as usize) == self.first.len()
+					|| i < actions.get_trailed_int(self.first[k as usize])
+				{
+					return Ok(());
+				}
+				lim = self.get_upper_limit(actions, k as usize);
+			}
+			i += 1;
+		}
+
+		if (i as usize) < self.vars.len() {
+			actions.set_int_lower_bound(self.vars[i as usize - 1], k, |a: &mut P| {
+				self.explain_lower(a, i as usize - 1, k)
+			})?;
+		}
+
+		let _ = actions.set_trailed_int(self.first[k as usize], 0);
+		Ok(())
+	}
+}
+
+impl<P, E> Propagator<P, E> for IntSeqPrecedeChainBounds
+where
+	P: PropagationActions,
+	E: ExplanationActions,
+{
+	#[tracing::instrument(name = "seq_precede_chain", level = "trace", skip(self, actions))]
+	fn propagate(&mut self, actions: &mut P) -> Result<(), Conflict> {
+		if self.initialized {
```

Comment:
I think I would like it better the other way around. Have a `if !self.initialized { return self.propagate_full(actions); }`, then the rest is more straightforward.